crates/xmtp-networking/src/grpc_api_helper.rs

```diff
@@ -35,23 +117,46 @@ impl Client {
             .parse()
             .map_err(|e| tonic::Status::new(tonic::Code::Internal, format!("{}", e)))?;
 
-        let mut client =
-            MessageApiClient::with_interceptor(self.channel.clone(), move |mut req: Request<()>| {
-                req.metadata_mut().insert("authorization", token.clone());
-                Ok(req)
-            });
-
-        return client
-            .publish(PublishRequest { envelopes })
-            .await
-            .map(|r| r.into_inner());
+        match &mut self.conn {
+            InnerApiConnection::Plain(c) => {
+                return MessageApiClient::with_interceptor(c, move |mut req: Request<()>| {
```

Comment:
You already figured out the interceptor approach, but I struggled with `with_interceptor` here. It didn't want to take the client that I was passing in. I had a realization that we can modify headers on the tonic request directly:
```
        let mut tonic_request = Request::new(request);
        // Insert the auth token into the request
        tonic_request.metadata_mut().insert("authorization", token);
        non_tls_client.publish(tonic_request).await
```

The longer story is the MessageApiClient has two components, a connector (TCP connection) and a part that handles the Body, in this case a Request. Rather than muck with an interceptor to modify the body, we can just modify it ourselves. Took me embarrassingly long to figure that out.

Reply:
https://github.com/xmtp/libxmtp/pull/62/commits/d9c994de9cd6746d467341bd769dd2acf159871a

Reply:
nice!